### DISCUSSION TITLE
What does `Mdi` in `Toplevel` really mean?

### DISCUSSION THREAD

**Discussion by tig** [tui-cs/Terminal.Gui#2357] [Q&A]
For @bdisp as I believe he has the best knowledge of all this...

Is the "Mdi" concept, as currently implemented in Terminal.Gui, really about "Mulitple Document Interface" or, is it really about "non-modal views that have z-order and can overlap"? 

Does "Not Mdi" means "non-modal views that have the same-zorder and are tiled"?

Also, do we know of any use-cases of Terminal.Gui's `Mdi` functionality outside of the `BackgroundWorkerCollection` scenario and unit tests? I've searched our codebase and github code-search and I can't find any uses of it in the wild. 

It appears the choice calling what we have in Terminal.Gui today for overlapping windows `Mdi` was driven by the fact that `WinForms` has such a concept. Is this correct?

I ask, because I have a POC of a refactor of `View`, `TopLevel`, etc... for v2 that supports non-modal, overlapping views that are movable/sizable, and none of the `Mdi` complexity is needed. I want to make sure I'm not missing anything.

**Comment by BDisp** [comment]
> For @BDisp as I believe he has the best knowledge of all this...
> 
> Is the "Mdi" concept, as currently implemented in Terminal.Gui, really about "Mulitple Document Interface" or, is it really about "non-modal views that have z-order and can overlap"?
> 
If you examine this code it's really about "non-modal views that have z-order and are tiled/overlap with a `MdiTop` which is the main window. Modal are show over all the others and hanging all the rest until they are closed.
```cs
public static List<Toplevel> MdiChildes {
	get {
		if (MdiTop != null) {
			List<Toplevel> mdiChildes = new List<Toplevel> ();
			foreach (var top in toplevels) {
				if (top != MdiTop && !top.Modal) {
					mdiChildes.Add (top);
				}
			}
			return mdiChildes;
		}
		return null;
	}
}
```
> Does "Not Mdi" means "non-modal views that have the same-zorder and are tiled"?
> 
"Not Mdi" (case of `IsMdiContainer` is false) means "non-modal views that only the Application.Top is shown. Only modal can be open over them.

> Also, do we know of any use-cases of Terminal.Gui's `Mdi` functionality outside of the `BackgroundWorkerCollection` scenario and unit tests? I've searched our codebase and github code-search and I can't find any uses of it in the wild.
> 
I really don't know, but as they are a little complex to develop may be it's not used. I think they are useful for dealing with various opens windows. They can be used for developing an IDE with text editors, properties, toolbox, etc...

> It appears the choice calling what we have in Terminal.Gui today for overlapping windows `Mdi` was driven by the fact that `WinForms` has such a concept. Is this correct?
> 
Correct.

> I ask, because I have a POC of a refactor of `View`, `TopLevel`, etc... for v2 that supports non-modal, overlapping views that are movable/sizable, and none of the `Mdi` complexity is needed. I want to make sure I'm not missing anything.

I think you don't missing anything. All the necessary code for multiple non-modal toplevels are already implemented, but I limited it to only `Mdi` concept. So, removing the Mdi and make it working as multiple non-modal toplevels by default works fine as well. But if you want to use some events that only works with Mdi, then you have to remove all the Mdi references, or if you don't want to use none of them you can remove but you'll have a bunch of unit tests errors :-)
